astropy/modeling/tests/test_model_sets.py

```diff
@@ -0,0 +1,226 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+"""
+This module tests model set evaluation for some common use cases.
+"""
+import pytest
+import numpy as np
+from numpy.testing.utils import assert_allclose
+
+from ..models import Polynomial1D, Polynomial2D
+from ..fitting import LinearLSQFitter
+from ..core import Model
+from ..parameters import Parameter
+
+
+x = np.arange(4)
+xx = np.array([x, x + 10])
+xxx = np.arange(24).reshape((3, 4, 2))
+
+
+class TParModel(Model):
+    """
+    A toy model to test parameters machinery
+    """
+    #standard_broadasting = False
+    inputs = ('x',)
+    outputs = ('x',)
+    coeff = Parameter()
+    e = Parameter()
+
+    def __init__(self, coeff, e, **kwargs):
+        super().__init__(coeff=coeff, e=e, **kwargs)
+
+    @staticmethod
+    def evaluate(x, coeff, e):
+        return x*coeff + e
+
+
+def test_model_axis_1():
+    """
+    Test that a model initialized with model_set_axis=1
+    can be evaluated with model_set_axis=False.
+    """
+
+    p1 = Polynomial1D(1, n_models=2, model_set_axis=1)
+    p1.c0 = [2, 3]
+    p1.c1 = [1, 2]
+    t1 = Polynomial1D(1, c0=2, c1=1)
+    t2 = Polynomial1D(1, c0=3, c1=2)
+
+    with pytest.raises(ValueError):
+        p1(x)
+
+    with pytest.raises(ValueError):
+        p1(xx)
+
+    y = p1(x, model_set_axis=False)
+    assert len(y) == 2
+    assert_allclose(y[0], t1(x))
+    assert_allclose(y[1], t2(x))
+
+    y = p1(xx, model_set_axis=False)
+    assert len(y) == 2
+    assert_allclose(y[0], t1(xx))
+    assert_allclose(y[1], t2(xx))
+
+    y = p1(xxx, model_set_axis=False)
+    assert_allclose(y[0], t1(xxx))
+    assert_allclose(y[1], t2(xxx))
+    assert len(y) == 2
+
+
+def test_model_axis_2():
+    """
+    Test that a model initialized with model_set_axis=2
+    can be evaluated with model_set_axis=False.
+    """
+    p1 = Polynomial1D(1, n_models=3, model_set_axis=2)
+    p1.c0 = [1, 2, 3]
+    p1.c1 = [10, 20, 30]
+    t1 = Polynomial1D(1, c0=1, c1=10)
+    t2 = Polynomial1D(1, c0=2, c1=20)
+    t3 = Polynomial1D(1, c0=3, c1=30)
+
+    with pytest.raises(ValueError):
+        p1(x)
+
+    with pytest.raises(ValueError):
+        p1(xx)
+
+    y = p1(x, model_set_axis=False)
+    assert y.shape == (4, 1, 3)
```

Comment:
I understand why this happens (I think), still, I wonder if it shouldn't be an error to evaluate a model with parameters having a shape smaller than implied by model_set_axis. Is there a good case for permitting it? It's seems like one more confusing element to dealing with model sets.

Reply:
This now is some pre-existing magic with polynomial models.
```
p = Polynomial1D(1, n_models=3, model_set_axis=2)
p1.c0.value
Out[54]: array([[[ 0.,  0.,  0.]]])

p1.c0 = [1, 2, 3]
p1.c0.value
Out[56]: array([[[ 1.,  2.,  3.]]])
```
It's not so in general:
```
g1= Gaussian1D(n_models=2, model_set_axis=2)

InputParameterError: All parameter values must be arrays of dimension at least 3 for 
model_set_axis=2 (the value given for 'amplitude' is only 0-dimensional)
```

Reply:
Furthermore, although the above polynomial example works, passing a negative axis fails:
```
p= Polynomial1D(1, n_models=3, model_set_axis=-1)

AxisError: axis -1 is out of bounds for array of dimension 0
```

Reply:
OK, but I wonder if there should be a test that relies on that. I doubt it.

Reply:
You may be right, though it does exist and changing it is a change in behavior. The problem is that some things work, some don't and how this was intended to work is not quite clear to me.
It's certainly not how I would expect it to work.

Reply:
My point is whether we should embed in tests behavior we consider dubious. Yes, if we eliminated it, it would be a change in behavior, but it probably is a good thing (I could point out a few more dubious behaviors that exist, but I don't want to enshrine them with tests :-)

Reply:
No, I get your point.
Would you agree that model sets need a rethink, clear definition and rewrite?
The goal would be to remove some of the extra functionality which works in some cases and not in others and remove ambiguity from code and interface.

Same question to @jehturner or anyone interested?

Reply:
I think for the purposes of solving the current problem in the short run, we can't deal with all these quirks (yes, I do think that they need a rethink and redo, but not immediately). Mainly just change this test not to depend on this particular behavior of filling in missing dimensions.

Reply:
You mean explicitly set the parameters to 

```
p1.c0 = [[[1, 2, 3]]]
p1.c1 = [[[10, 20, 30]]]
```
Otherwise I think the line below is correct.
```
assert y.shape == (4, 1, 3)
```

Reply:
Yeah, that would do it.

Reply:
I think +1; this behaviour is inconsistent with instantiation, where the same parameters have to be 3-dimensional, right?

This is a pretty odd test case in the first place -- a `Polynomial1D` with `model_set_axis=2` -- but I suppose why not (one could be doing something like supersampling a spectrum using data from multiple image rows)?

Yes, some clarification, simplification and greater consistency of model set behaviour (& other features of modeling) would be great -- but it's not 100% clear to me what features are really non-critical for other people, nor whether we still have the effort for a proper rewrite (have you?). I'm risking being seen as a bit off-base by fixing numerous things that are peripherally-related to what I'm eventually trying to achieve and will get sucked into some other work for a while next month. I'm happy to provide input, but I'm not sure how much of a dent I could put in overhauling the core code, personally... But interested, certainly :wink:.